Providing one cmd_out per timestamp.

joaobrittoneto opened the pull request:
Every step a sample of cmd_out is output even in case there is no new
data in cmd_in or pose_samples.
The timestamp of cmd_out is defined by cmd_in

joaobrittoneto committed:
Test that shows the repeated timestamp in the pid control's cmd_out

Every step a sample of cmd_out is output even in case there is no new
data in cmd_in or pose_samples.
The timestamp of cmd_out is defined by cmd_in

joaobrittoneto committed:
Make the pose_samples drive the PIDController

A new cmd_out is sent only when there is a new pose_sample.
The timestamp of the cmd_out is the same of the pose_sample.
It happens cause the cmd_in, being the set-point of the controller,
doesn't need to be received all time. In the other handy, the pose_sample
being the feedback of the system need to be compared constantly in order
to generate a new cmd_out.

doudou commented on tasks/PIDController.cpp:
```diff
@@ -44,21 +44,18 @@ bool PIDController::startHook()
+    {
+        if(new_pose_samples_timeout.elapsed())
+        {
+            exception(POSE_TIMEOUT);
```
:+1:

saarnold replied:
I'm fine with separating the states to a soft and a hard reference sample timeout.
But we should adapt the other components as well in order to have an equal behavior.

joaobrittoneto replied:
Ok

doudou commented:
you probably should rename the PR

saarnold commented:
So this PR changes the control chain in a way that the frequency of the reference samples will determine the frequency of the output samples and not the periodicity of the tasks or the frequency of the input commands. Is this what we want? I'm honestly not sure.
One thing I see: The reference providing component might not be able to provide samples in a constant or high enough frequency, so the components and drivers receiving the commands from the control chain might run into timeouts.

The reason why a sequence of commands right now might not change their timestamp or values is to be able to provide a constant frequency of commands. One solution to avoid that would therefore be to reduce the periodicity of the chain.
The question is if receiving a gap is better or worse than receiving the same command multiple times?

doudou commented:
> The question is if receiving a gap is better or worse than receiving the same command multiple times?
> the components and drivers receiving the commands from the control chain might run into timeouts.

The PID controllers are closed-loop controllers. If you don't close the loop often enough, you're seriously screwed. If you feed a long-latency estimator to a close-loop controller, but update its command, you're really doing open-loop control - which is definitely not good. They _should_ run into timeouts if they don't get a state estimate often enough !

Moreover, using the state estimate as master trigger does not preclude from timing out on the input commands as well for safety.

joaobrittoneto commented:
I do understand your concerns.
But the repeated timestamp doesn't seems right to me. In my opinion, a command, or sample, should be a function of time, it means that for one timestamp there is one and only one ouput.
One could just add the task's step in the timestamp to make it a function of time (but the hand tunning of data may be questionable). 

Regarding the functionality of the PIDController, if I understand it right, the `cmd_in` is the set-point of the controller, the desired state, while the `pose_samples` is the feedback of the controller, the measurement (or estimation) of the actual state. From my point of view, there is no need to receiving a `cmd_in` constantly. Once there is a new set-point (`cmd_in`) the controller should provide a new `cmd_out` according the measurement of the actual state (`pose_samples`). If the controller doesn't know the actual state, why it should provide a new command?

About the frequency of `cmd_out` effecting the behavior of the actuator, I think it should not be handle by the PIDController, but by other component. I mean, the enitech's thruster has as input a set-point of rotor speedy, so no need to providing the same cmd (set-point) every period.
If the actuator requires a cmd at fix rate for operate (an open-loop actuator), then an end-chain task could keep the periodicity of the cmd, like replicating the last cmd every step.

saarnold commented:
> They should run into timeouts if they don't get a state estimate often enough!

Of course. I was thinking of an actuator which is not controlling at all or not correctly because of a too low frequency. But yes, that could also be easily solved by the driver.

> Regarding the functionality of the PIDController, if I understand it right, the cmd_in is the set-point of the controller, the desired state, while the pose_samples is the feedback of the controller, the measurement (or estimation) of the actual state. From my point of view, there is no need to receiving a cmd_in constantly. Once there is a new set-point (cmd_in) the controller should provide a new cmd_out according the measurement of the actual state (pose_samples). If the controller doesn't know the actual state, why it should provide a new command?

Ok sounds reasonable, other than that I also can't think of any other reason right now why it might was implemented that way.

joaobrittoneto committed:
Make `ruby test/test_base.rb` working

Update require to `minitest/autorun` and use a WorldToAligned to test the
functionality.
Need to remove the describes inside the test. It was not identifying the
task (??)
Some tests are failing. To be fixed.

joaobrittoneto committed:
Remove state WAIT_FOR_POSE_SAMPLE for BasePIDController

Once the controller will only provide a command on the arrival of a
new pose_sample, avoid to switch between CONTROLLING and
WAIT_FOR_POSE_SAMPLE states constantly. Just keep a POSE_TIMEOUT exception
for long delays of pose.

joaobrittoneto committed:
Create POSE_TIMEOUT exception for WorldToAligned

Create also ORIENTATION_TIMEOUT exception for AlignedToBody.

Define WAIT_FOR_POSE/ORIENTATION_SAMPLE as runtime states,
once the perspectives tasks wait for the first data in that state.
Once it's an OldData for pose, the tasks keep in CONTROLLING
(using the last pose) until it reach the TIMEOUT exception.

As both tasks do a frame transformation between gatheredInputCommand
to cmd_out, there is only a cmd_out for a NEW_COMMAND in gatherInputCommand

joaobrittoneto committed:
Refactor Base

Create OneLoop that embedded gatherInputCommand and calcOutPut.
OneLoop is called in both Base::runtimeHook and Base::errorHook.
The State transition is done in both Hook, according the state returned by
OneLoop.
In case the derived class should switch to its own error/runtime state,
it should be done in the re-implementation of calcOutput and return false.

Create enum CommandStatus and make gatherInputCommand assign both
the command and its status in LinearAngular6DCommandStatus. Make them be
the argument for calcOutput.

joaobrittoneto commented:
@saarnold, @doudou 
I tried to embedded the same behavior of PIDController in the other tasks, but I could not do it easily. 
It was really confusing for me the Base and the state transitions in gatherInputCommand.
So I made some code refactoring in order to make the code more clear.

joaobrittoneto committed:
Define specific pose & orientation timeouts

Make AlignedToBody and WorldToAligned use specific timeout for pose and
orientation.
Update test/test_base.rb for TIMEOUT error and recovery case of cmd_in

joaobrittoneto committed:
Include test for PIDController and validate output of cmd_out

joaobrittoneto committed:
Code-refactoring in BasePIDController and PIDController

Create structs for putting together linear and angular PIDs, currentState
and currentCov.
Embedded timestamp in LinearAngular6DPIDState, once it's present as
an output_port.

Define specific timeout_pose in BasePIDController and create
calcPIDStateCommand. It's called in calcOutput and requires
reference (merged_command), the current_state (pose_sample) and the PIDs
(they are updated). It provides the command_output and the pid_state and
their timestamp are set from the current_state.

doudou commented on auv_control.orogen:
```diff
@@ -96,6 +96,10 @@ task_context 'BasePIDController' do
 
+    # This property defines the timeout for the pose_samples input port in seconds.
+    # 0 means that the timeout would be ignored.
+    property "timeout_pose", "double", 1
```
use `base::Time` for times.

joaobrittoneto replied:
Ok.
Since we are here, how about the Base's [timeout_in](https://github.com/rock-control/control-orogen-auv_control/blob/master/auv_control.orogen#L38) and `addCommandInput`'s [timeout](https://github.com/rock-control/control-orogen-auv_control/blob/master/auv_control.orogen#L65)?

joaobrittoneto replied:
@doudou 
ping.

doudou commented on auv_control.orogen:
```diff
@@ -67,7 +67,7 @@ task_context "Base" do
 
     runtime_states :CONTROLLING, :CONTROLLING_UNSAFE
-    error_states :INPUT_MISSING, :INPUT_COLLIDING, :INPUT_UNEXPECTED, :TIMEOUT, :WAIT_FOR_CONNECTED_INPUT_PORT, :WAIT_FOR_INPUT 
+    error_states :INPUT_MISSING, :INPUT_COLLIDING, :INPUT_UNEXPECTED, :TIMEOUT, :WAIT_FOR_CONNECTED_INPUT_PORT, :WAIT_FOR_INPUT, :ERROR_CALCOUTPUT
```
Don't create generic error states like `ERROR_CALCOUTPUT`. If you need one, you can simply transition to the generic RTT error state.

doudou commented on tasks/Base.cpp:
```diff
@@ -114,6 +109,27 @@ void Base::cleanupHook()
     }
 }
 
+Base::States Base::OneLoop()
```
Naming convention. Either `oneLoop()` or `performOneLoop()`

doudou commented on tasks/Base.cpp:
```diff
@@ -241,33 +240,46 @@ bool Base::addCommandInput(std::string const & name, double timeout){
+    return CONTROLLING;
 }
 
+std::vector<Base::InputPortInfo*> Base::checkConnectedPorts(std::vector<InputPortInfo> &in_ports) const
```
I know it's not a hard requirement, but I would prefer that you avoid allocating on the heap during the updateHook. Keep an attribute in the class of type `std::vector<Base::InputPortInfo*>` that would be e.g. `connected_input_ports` and reserve() its size once and for all. Then, pass it as argument to this method.

joaobrittoneto committed:
Replace timeout_pose and timeout_orientation from double to base::Time

Remove generic and unused error state ERROR_CALCOUTPUT.
Rename performOneLoop() from OneLoop()

joaobrittoneto committed:
Replace calcOutput() by calcOutput(const LinearAngular6DCommandStatus &cmd)

calcOutput receives command and its status as argument.
Remove merged_command as global attribute.

joaobrittoneto committed:
Define attribute connected_input_ports

Reserve its size to 6 in configureHook. Pass it as argument in
checkConnectedPorts.
Update test_base.rb to use timeout as base::Time

joaobrittoneto committed:
Make AlignedToBody not output cmd_out of old command

joaobrittoneto committed:
Replace defined enum CommandStatus by RTT::FlowStatus

LinearAngular6DCommandStatus carries not only the command as well the
status in RTT::FlowStatus standard.
This struct is passed to calcOutput() so the derived class can define
what to do with the command knowing it status.

joaobrittoneto committed:
gatherInputCommand only set merged_command.status as new or old

The status of the command is only valid (or used) when the return of
gatherInputCommand is CONTROLLING.
If it returns an error state, calcOutput is not called and the
merged_command.status is irrelevant.

saarnold commented on tasks/Base.hpp:
```diff
@@ -24,34 +24,73 @@ namespace auv_control {
+        /** Gather input from connected input ports
+         *
+         * @param merging_command, where the commands will be safe
+         * @param cmd_status, Define the status of merging_command, NEW_COMMAND, OLD_COMMAND, NO_COMMAND or INVALID_COMMAND
```
This parameter doesn't seem to be existent any more.

joaobrittoneto replied:
Oh yeah
:+1:

saarnold commented on tasks/Base.cpp:
```diff
@@ -241,33 +233,44 @@ bool Base::addCommandInput(std::string const & name, double timeout){
+bool Base::checkConnectedPorts(std::vector<Base::InputPortInfo> &input_ports, std::vector<Base::InputPortInfo*> &connected_input_ports)
+{   // Clear vector before assign connected ports
+    if(!connected_input_ports.empty())
+        connected_input_ports.erase(connected_input_ports.begin(), connected_input_ports.end());
```
Would calling clear here not do the same trick?

joaobrittoneto replied:
Indeed

saarnold commented:
Looks good to me :+1:

joaobrittoneto committed:
Update tests for timeout_pose

Update documentation on gatherInputCommand.
Use clear instead of erase in checkConnectedPorts

doudou commented on 6dControl.hpp:
```diff
@@ -27,8 +28,25 @@ namespace auv_control{
+        motor_controller::PID angular[3];
+    };
+
+    struct LinearAngular6DCommandStatus
```
This header is parsed / generated by orogen, stick to public interface types. Put private structs somewhere else.

doudou commented on auv_control.orogen:
```diff
@@ -96,6 +96,10 @@ task_context 'BasePIDController' do
 
+    # This property defines the timeout for the pose_samples input port.
+    # 0 means that the timeout would be ignored. Default is 1 second
+    property "timeout_pose", "base::Time"
```
Use the typelib /base/Time notation, it's closer to the rest of the file.
(Same for the rest of the file)

doudou commented:
Looks good to me too.

joaobrittoneto committed:
Move LinearAngular6DCommandStatus from 6DControl.hpp to task/Base.hpp

Use notation "/namespace/struct" instead of "namespace::struct" in .orogen

joaobrittoneto commented:
@doudou 
Any other comment? 
The timeouts in Base are still as [double](https://github.com/joaobrittoneto/control-orogen-auv_control/blob/88d0c776206b1289062f9b02a78e6cc09dc42c3b/auv_control.orogen#L38). Should they also be shifted to base/Time?

doudou commented:
> The timeouts in Base are still as double. Should they also be shifted to base/Time?

That would be one comment. Other than that, as I said, 

> > Looks good to me too.

joaobrittoneto committed:
Change Base's timout from double to base/Time

Update addCommandInput, Base, tests and config files

joaobrittoneto commented:
@saarnold,
Do you see any backward compatibility issue with the last commit?
Otherwise it's ready to merge.

saarnold commented:
> Do you see any backward compatibility issue with the last commit?

Well changing the type of the timeout property will obviously break backward compatibility.
But that's fine, we just announce that this PR will be merged by the end of this week.
Thanks for the effort!
